File: backend/app/services/delivery_service.py
from datetime import datetime
from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.repositories.delivery_repo import delivery_repo
from app.repositories.order_repo import order_repo
from app.schemas.delivery import DeliveryUpdate
from app.models.user import UserRole, User
from app.models.delivery import Delivery

class DeliveryService:
    async def claim_delivery(self, db: AsyncSession, delivery_id: str, current_user: User) -> Delivery:
        """Assign an active shipment task to a claiming Delivery Partner."""
        if current_user.role != UserRole.DELIVERY_PARTNER.value and not current_user.is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only authenticated Delivery Partners can claim a shipment."
            )

        delivery = await delivery_repo.get(db, delivery_id)
        if not delivery:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Delivery record not found."
            )

        if delivery.partner_id is not None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="This delivery assignment has already been claimed."
            )

        # Claim the delivery
        delivery.partner_id = current_user.id
        delivery.status = "picked_up"
        delivery.updated_at = datetime.utcnow()
        
        # Sync the parent order status
        order = await order_repo.get(db, delivery.order_id)
        if order:
            order.status = "Shipped"

        db.add(delivery)
        await db.commit()
        await db.refresh(delivery)
        return delivery

    async def update_delivery_status(self, db: AsyncSession, delivery_id: str, update_in: DeliveryUpdate, current_user: User) -> Delivery:
        """Update route tracking states and estimate completions."""
        delivery = await delivery_repo.get(db, delivery_id)
        if not delivery:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Delivery record not found."
            )

        # Enforce carrier ownership verification
        if delivery.partner_id != current_user.id and not current_user.is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not authorized to update this shipment."
            )

        if update_in.status:
            delivery.status = update_in.status
            if update_in.status == "delivered":
                delivery.actual_delivery = datetime.utcnow()
                # Update base order status
                order = await order_repo.get(db, delivery.order_id)
                if order:
                    order.status = "Delivered"

        delivery.updated_at = datetime.utcnow()
        db.add(delivery)
        await db.commit()
        await db.refresh(delivery)
        return delivery

delivery_service = DeliveryService()

File: backend/src/services/deliveryService.js
import supabase from '../db/supabase.js';
import { AppError } from '../middleware/error.js';

export const deliveryService = {
  async claimDelivery(deliveryId, currentUser) {
    if (currentUser.role !== 'delivery_partner' && !currentUser.is_admin) {
      throw new AppError(403, 'Only authenticated Delivery Partners can claim a shipment.');
    }

    // 1. Fetch delivery
    const { data: delivery, error } = await supabase
      .from('deliveries')
      .select('*')
      .eq('id', deliveryId)
      .single();

    if (error || !delivery) {
      throw new AppError(404, 'Delivery record not found.');
    }

    if (delivery.partner_id !== null) {
      throw new AppError(400, 'This delivery assignment has already been claimed.');
    }

    // 2. Claim delivery
    const { data: updatedDelivery, error: updateError } = await supabase
      .from('deliveries')
      .update({
        partner_id: currentUser.id,
        status: 'picked_up',
        updated_at: new Date().toISOString()
      })
      .eq('id', deliveryId)
      .select('*')
      .single();

    if (updateError) {
      throw new AppError(500, `Error claiming delivery: ${updateError.message}`);
    }

    // 3. Sync parent order status to Shipped
    await supabase
      .from('orders')
      .update({ status: 'Shipped' })
      .eq('id', delivery.order_id);

    return updatedDelivery;
  },

  async updateDeliveryStatus(deliveryId, status, currentUser) {
    // 1. Fetch delivery
    const { data: delivery, error } = await supabase
      .from('deliveries')
      .select('*')
      .eq('id', deliveryId)
      .single();

    if (error || !delivery) {
      throw new AppError(404, 'Delivery record not found.');
    }

    // 2. Verify ownership
    if (delivery.partner_id !== currentUser.id && !currentUser.is_admin) {
      throw new AppError(403, 'You are not authorized to update this shipment.');
    }

    const updates = {
      status,
      updated_at: new Date().toISOString()
    };

    if (status === 'delivered') {
      updates.actual_delivery = new Date().toISOString();
    }

    // 3. Update delivery
    const { data: updatedDelivery, error: updateError } = await supabase
      .from('deliveries')
      .update(updates)
      .eq('id', deliveryId)
      .select('*')
      .single();

    if (updateError) {
      throw new AppError(500, `Error updating delivery status: ${updateError.message}`);
    }

    // 4. Update parent order status to Delivered
    if (status === 'delivered') {
      await supabase
        .from('orders')
        .update({ status: 'Delivered' })
        .eq('id', delivery.order_id);
    }

    return updatedDelivery;
  }
};

export default deliveryService;
